### PR TITLE
Fast Clone Deletion

### PR DESCRIPTION
**Pull request by shartse**
[Video of talk](https://www.youtube.com/watch?v=yQGGwhDxkdw&t=179s) at BSDCan

**Note:** I've updated the deletion performance figures and test script since this presentation as they weren't doing quite what I thought they were. Effectively, I was passing the `stride` option in `dd` in bytes instead of blocks. This meant I was writing more than I thought to the clone and in the sparse case I was writing off the end of the file. The old tests still demonstrated meaningful performance differences, just in a less realistic scenario. 

## Motivation and Context

Deleting zfs clones requires determining which parts of the block tree belong to the clone and which are still part of the original snapshot. The basic algorithm is to traverse the block tree, at each block checking when it was born. If it was created after the clone was created then free it and recursively check its child blocks. If the block was born before the clone was created then we know it is part of the snapshot. Otherwise, continue traversing the block tree. 

This algorithm is non ideal in the case of sparsely modified clones (those with a small number of scattered writes). Even though there is very little written to the clones it takes a long time to scan through the block tree, i.e. deletion time for a clone with sparse writes is proportional to the size of the underlying block tree, which is the size snapshot. Our goal to change this be to proportional to the space used by the clone only. 
 
The following figure shows the performance difference between deleting clones with with the same amount of modification, just with contiguous vs. sparsely distributed writes. 

deletion time of sparse vs contiguously written clones

The script I wrote for this test is available here: [lx_fast_delete](https://github.com/zfsonlinux/zfs/files/2867114/lx_fast_delete.txt)

### Livelist Algorithm 
Livelists use the same `deadlist_t` struct as Deadlists and are also used to track block pointers over the lifetime of a dataset. Livelists however, belong to clones and track the block pointers that are clone-specific (were born after the clone's creation). The exception is embedded block pointers which are not included in Livelists because they do not need to be freed.

When it comes time to delete the clone, the livelist provides a quick reference as to what needs to be freed. For this reason, livelists also track when clone-specific block pointers are freed before deletion to prevent double frees. Each block pointers in a livelist is marked as a `FREE` or an` ALLOC` and the deletion algorithm iterates backwards over the livelist, matching `FREE`/`ALLOC` pairs and then freeing those `ALLOCs` which remain. Livelists are also updated in the case when block pointers are remapped during removal: the old version of the blkptr is cancelled out with a `FREE` and the new version is tracked with an `ALLOC`.

#### Sublists

To bound the amount of memory required for deletion, livelists over a certain size are spread over multiple entries (made natural by the use of the `deadlist_t` data structure). . Entries are grouped by birth txg so we can be sure the `ALLOC`/`FREE` pair for a given block pointer will be in the same entry. This allows us to delete livelists incrementally over multiple syncs, one entry at a time. The threshold size at which we create a new sub-livelist is an important tunable for livelist performance - we had to balance the fact that larger sublists mean fewer sublists (decreasing the cost of insertion) against the consideration that sublists will be loaded into memory and shouldn't take up an inordinate amount of space. We settled on ~500000 entries, corresponding to roughly 128M

#### Condensing
We still have the issue that the livelist can grow arbitrarily large - continually gaining sub-livelists. However, because of the restriction that block pointers of the same txg must reside in the same sub-livelist, we can determine if a given block has been freed (in time proportional to the size of the sub-livelist instead of the total number of frees and allocs). This means we can periodically clear out `ALLOC`/`FREE` pairs from sublists and then merge the smaller lists together.

#### Disabling

Livelists are not always the most effective deletion method. We can approximate how much of a performance gain a livelist will give us based on the pecentage of blocks shared between the clone and its origin. Zero percent shared means that the clone has completely diverged and that the old method is maximally effective: every read from the block tree will result in lots of frees. Livelists give us gains when they track blocks scattered across the tree, when one read in the old method might only result in a few frees. Once the clone has been overwritten enough, writes are no longer sparse and we'll no longer get much of a benefit from tracking them with a livelist. We chose a lower limit of 75 percent shared (25 percent overwritten). Once the amount of shared space drops below this threshold, the clone will revert to the old deletion method.

#### Deletion
deletion time of sparsely written clones

I repeated the first test discussed above for deleting sparsely written clones and found a dramatic performance improvement using the new algorithm. As the above figure shows, the old deletion time grows linearly with respect to the size of the snapshot, while the new deletion time grows at a much smaller rate. In the following figure, I compare the new deletion time to the size of the data written to the clone and we can see that it's now proportional to that. 

new algorithm vs clone size

I also examined performance changes to the old algorithm's "best case" scenario - that of contiguously overwritten clones. I wasn't expecting to be a performance improvement with the livelist strategy since the old method was already very efficient (because of the way the overwritten block pointers laid out in the block tree) and we can see from the figure below that the livelist does add some overhead. However, since the deletion time is still proportional to the amount of data written and real world scenarios are likely to be a hybrid of sparse and contiguous write patterns, I'm not concerned about this performance change.  

deletion time of contiguously written clones

#### Writes

One of the main drawbacks of the livelist deletion algorithm is the way that it imposes extra work done at write-time to track the modified block pointers for clones. I explored possible write performance degradation by constructing a "worst case" scenario for writes. This scenario is one in which a write modifying blocks tracked in many different livelist sublists and a `FREE` entry has to be appended to each one. 

1. Create a snapshot of a large filesystem and a clone of that snapshot
2. Write `X` bytes of data in `N` different chunks to the clone such that each chunk is separated into it's own sublist.
3. Overwrite the modified region with a small write `Y` but spread evenly into `N` piece such that each touches a different sublist. 
4. Measure the time the write in 3 took, including the `sync` time which is when the block pointers are flushed to the livelist. 

The script I used for this test is available here: [lx_write_test](https://github.com/zfsonlinux/zfs/files/2870698/lx_write_test.txt)

The following figure shows the results from repeated experiments with `N = 50`, `N = 1` and `N = 0` (the old algorithm that doesn't use livelists). We can see that a larger number of sublists does result in degraded write performance. This motivates the reasoning for having the sublist size be as large as is reasonable, given memory constraints, as well as the condensing of sublists and disabling of livelists altogether once the clone has been overwritten past a certain point. 

writes

Note: the figure shows the average time of 1000 writes and the standard deviation. 

### Correctness
I added zfs tests to cover basic assumptions about livelists and well as edge cases for deletion, condensing and disabling. 

This *Illumos* implementation feature has been active internally at Delphix for the past 10 months.

I've been running these changes through `zloop` for several days without hitting anything related to this. 

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/zfs/zthr.c [hunk 1]
```diff
@@ -447,3 +456,75 @@ zthr_iscancelled(zthr_t *t)
...
+	 *     the behavior is similar to [3].
+	 * [5] The thread is the middle of being cancelled, which is
+	 *     similar to [3]. We'll wait for the cancel, which is
+	 *     waiting for the zthr func.
+	 *
+	 * Since requests are serialized, by the time that we get
+	 * control back we expect that the zthr has completed it's
+	 * zthr_func.
+	 */
+
```
Comment (sdimitro) [inline module/zfs/zthr.c]: (nit) empty line

File: module/zfs/zthr.c [hunk 2]
```diff
@@ -207,6 +207,9 @@ struct zthr {
 	/* flag set to true if we are canceling the zthr */
 	boolean_t	zthr_cancel;
 
+	/* flag set to true if we are waiting for the zthr to finish */
+	boolean_t	zthr_finish;
+	kcondvar_t	zthr_finish_cv;
```
Comment (sdimitro) [inline module/zfs/zthr.c]: I'm trying to remember why we decided to not name this some variation of `zthr_wait`.
If it is only the fact that we don't want to associate it with `zthr_wait_time` which is something different, then it may worth to change that variable name, and keep `zthr_wait` for our variables.

Maybe we can we can do the following changes:
`zthr_wait_time` -> `zthr_sleep_timeout`

`zthr_finish_cv` -> `zthr_wait_cv`/`zthr_waiter_cv`
`zthr_finish` -> `zthr_haswaiters`
`zthr_finish()` -> `zthr_wait_cycle_done()`
`zthr_isfinishing()` -> `zthr_has_waiters()`

The proposed names are not absolute, I'm just afraid that with the current code `finish` may be confused with `cancel`. Let me know what you think.

File: module/zcommon/zfeature_common.c [hunk 3]
```diff
@@ -348,6 +348,11 @@ zpool_feature_init(void)
 	    ZFEATURE_FLAG_MOS | ZFEATURE_FLAG_ACTIVATE_ON_ENABLE,
 	    ZFEATURE_TYPE_BOOLEAN, NULL);
 
+	zfeature_register(SPA_FEATURE_LIVELIST,
+	    "com.delphix:livelist", "livelist",
+	    "Improved clone deletion performance.",
```
Comment (ahrens) [inline module/zcommon/zfeature_common.c]: I think that this should depend on SPA_FEATURE_EXTENSIBLE_DATASET, since it adds DD_FIELD_LIVELIST.

File: tests/zfs-tests/include/libtest.shlib [hunk 4]
```diff
@@ -3547,6 +3547,24 @@ function mdb_get_uint32
 	return 0
 }
 
+#
+# Get the hexadecimal value of global variable using mdb.
+#
+function mdb_get_hex
```
Comment (sdimitro) [inline tests/zfs-tests/include/libtest.shlib]: We should take this mdb wrapper out 💔

File: module/zfs/dsl_deadlist.c [hunk 5]
```diff
@@ -566,3 +765,118 @@ dsl_deadlist_move_bpobj(dsl_deadlist_t *dl, bpobj_t *bpo, uint64_t mintxg,
...
+	    .to_free = to_free,
+	    .t = t
+	};
+	err = bpobj_iterate_nofree(bpobj, dsl_livelist_iterate, &arg);
+
+	avl_destroy(&avl);
+	return (err);
+}
+
+#if defined(_KERNEL)
```
Comment (sdimitro) [inline module/zfs/dsl_deadlist.c]: We should at these tunables to the zfs-parameters man page (especially `min_percent_shared` as it seems important for users outside of testing)

File: module/zfs/dsl_deadlist.c [hunk 6]
```diff
@@ -566,3 +765,118 @@ dsl_deadlist_move_bpobj(dsl_deadlist_t *dl, bpobj_t *bpo, uint64_t mintxg,
...
+/* CSTYLED */
+module_param(zfs_livelist_max_entries, ulong, 0644);
+MODULE_PARM_DESC(zfs_livelist_max_entries,
+	"Size to start the next sub-livelist in a livelist");
```
Comment (ahrens) [inline module/zfs/dsl_deadlist.c]: I believe (though @behlendorf could confirm) that we want all module parameters to be documented in the zfs-module-parameters.5 manpage.

File: module/zfs/bpobj.c [hunk 7]
```diff
@@ -263,22 +267,23 @@ propagate_space_reduction(bpobj_info_t *bpi, uint64_t freed,
...
-	uint64_t freed = 0, comp_freed = 0, uncomp_freed = 0;
+	int64_t freed = 0, comp_freed = 0, uncomp_freed = 0;
 	dmu_buf_t *dbuf = NULL;
 	bpobj_t *bpo = bpi->bpi_bpo;
 
-	for (int64_t i = bpo->bpo_phys->bpo_num_blkptrs - 1; i >= 0; i--) {
+	for (int64_t i = bpo->bpo_phys->bpo_num_blkptrs - 1; i >= start; i--) {
 		uint64_t offset = i * sizeof (blkptr_t);
 		uint64_t blkoff = P2PHASE(i, bpo->bpo_epb);
+		boolean_t bp_freed;
```
Comment (sdimitro) [inline module/zfs/bpobj.c]: (style nit)
inline declaration closer to the usage of the variable:
```
boolean_t bp_freed = BP_GET_FREE(bp);
err = func(arg, bp, bp_freed, tx);
```

File: module/zfs/bpobj.c [hunk 8]
```diff
@@ -263,22 +267,23 @@ propagate_space_reduction(bpobj_info_t *bpi, uint64_t freed,
 
 static int
 bpobj_iterate_blkptrs(bpobj_info_t *bpi, bpobj_itor_t func, void *arg,
-    dmu_tx_t *tx, boolean_t free)
+    int64_t start, dmu_tx_t *tx, boolean_t free)
```
Comment (sdimitro) [inline module/zfs/bpobj.c]: Looking back at the original internal review we had changed the `free` boolean argument here that naming in all the related functions from `free` to `delete` to make things more precise. Was there any reason why you didn't try to do that in this PR too? (e.g. is the change big enough that it could be in a separate PR? is the meaning of `free` more tightly coupled with other things here, than the original code in illumos? etc..)

File: module/zfs/dsl_dir.c [hunk 9]
```diff
@@ -2224,6 +2242,86 @@ dsl_dir_is_zapified(dsl_dir_t *dd)
...
+			 * by 'syncing' field of to_condense) and we can expect
+			 * it to clear to_condense on its own. Otherwise, we
+			 * returned before the zthr ran. The checkfunc will
+			 * now fail as cancelled == B_TRUE so we can safely NULL
+			 * out ds, allowing a different dir's livelist to be
+			 * condensed.
+			 *
+			 * We can be sure that the to_condense struct will not
+			 * be repopulated during the execution of this function
+			 * because dsl_livelist_try_condense excutes in syncing
```
Comment (sdimitro) [inline module/zfs/dsl_dir.c]: `s/excutes/executes`

File: module/zfs/dsl_dir.c [hunk 10]
```diff
@@ -2224,6 +2242,86 @@ dsl_dir_is_zapified(dsl_dir_t *dd)
...
+			 * context.
```
Comment (sdimitro) [inline module/zfs/dsl_dir.c]: `... because dsl_livelist_try_condense executes in syncing context.`
may be add at the end of this sentence
`... executes in syncing context, like this function.`
or something along those lines emphasizing your point more that both the current function and `dsl_livelist_try_condense()` are in syncing context, therefore they can't be run concurrently.

File: module/zfs/bpobj.c (hunk 8 above)
Comment (shartse) [reply module/zfs/bpobj.c]: So the reason I originally replaced `free` with `delete` was because I wanted to use the `free` variable to indicate the status of the `blkptr`. I decided to use `bp_freed` for that instead and leave `free` as it was to minimize code change in this area. I don't have strong feelings on whether `free` or `delete` is more precise but I'm leaning towards sticking with the existing convention.

File: module/zfs/zthr.c (hunk 2 above)
Comment (shartse) [reply module/zfs/zthr.c]: Yeah, I think that's a reasonable proposal. I think mostly I was dissatisfied with the ambiguity of the `zthr_wait*` names I had been using and want to go for something a little more specific. I think your suggestion addresses that and more accurately reflects that this is "waiting-type" behavior, not cancelling.

File: module/zfs/dsl_deadlist.c (hunk 6 above)
Comment (shartse) [reply module/zfs/dsl_deadlist.c]: Makes sense - this one in particular could definitely be used for performance tuning. 

I do have some tunables that I use exclusively in tests (i.e. `zfs_livelist_condense_zthr_pause`) and can't imagine any users wanting to use - basically variables to cause hangs and count race conditions. @behlendorf would you want those included in the man page as well or should I leave them out?

File: module/zfs/bpobj.c (hunk 8 above)
Comment (sdimitro) [reply module/zfs/bpobj.c]: That makes sense to me. Thank you for the clarification!

File: module/zfs/dsl_deadlist.c (hunk 5 above)
Comment (shartse) [reply module/zfs/dsl_deadlist.c]: I added entries for `zfs_livelist_max_entries` and `zfs_livelist_min_percent_shared`, the ones that impact performance. I'll leave out the others for now unless someone expresses a strong opinion about it.

File: module/zfs/dsl_deadlist.c (hunk 6 above)
Comment (shartse) [reply module/zfs/dsl_deadlist.c]: I added entries for `zfs_livelist_max_entries` and `zfs_livelist_min_percent_shared`, the ones that impact performance. I'll leave out the others for now unless someone expresses a strong opinion about it.

File: cmd/zdb/zdb.c [hunk 11]
```diff
@@ -518,7 +523,8 @@ dump_bpobj(objset_t *os, uint64_t object, void *data, size_t size)
 			(void) printf("got error %u from dmu_read\n", err);
 			break;
 		}
-		snprintf_blkptr_compact(blkbuf, sizeof (blkbuf), &bp);
+		boolean_t free = BP_GET_FREE(&bp);
+		snprintf_blkptr_compact(blkbuf, sizeof (blkbuf), &bp, free);
```
Comment (kithrup) [inline cmd/zdb/zdb.c]: There doesn't seem to be much point to having a separate variable there -- it's not used anywhere else, and it uses the name of a standard C function. I'd suggest just changing the function call to have the BP_GET_FREE(&bp) call inline.

File: cmd/zdb/zdb.c [hunk 12]
```diff
@@ -1404,14 +1410,19 @@ blkid2offset(const dnode_phys_t *dnp, const blkptr_t *bp,
 }
 
 static void
-snprintf_blkptr_compact(char *blkbuf, size_t buflen, const blkptr_t *bp)
+snprintf_blkptr_compact(char *blkbuf, size_t buflen, const blkptr_t *bp,
+    boolean_t free)
```
Comment (kithrup) [inline cmd/zdb/zdb.c]: Similarly, the name "free" is a bit problematic. show_free would be more descriptive, but a shorter version of the name would be fine.

File: cmd/zdb/zdb.c [hunk 13]
```diff
@@ -4209,6 +4309,100 @@ count_block_cb(void *arg, const blkptr_t *bp, dmu_tx_t *tx)
...
+static void
+dump_livelist_cb(dsl_deadlist_t *ll, void *arg)
+{
+	ASSERT3P(arg, ==, NULL);
+	dump_blkptr_list(ll, "Deleted Livelist");
+}
+
+/*
+ * Print out, register object references to, and increment feature counts for
+ * livelists that have been destroy by the user but haven't yet been freed.
```
Comment (kithrup) [inline cmd/zdb/zdb.c]: "destroyed". :)

File: module/zfs/bplist.c [hunk 14]
```diff
@@ -75,3 +75,17 @@ bplist_iterate(bplist_t *bpl, bplist_itor_t *func, void *arg, dmu_tx_t *tx)
...
 }
+
+void
+bplist_clear(bplist_t *bpl)
+{
+	bplist_entry_t *bpe;
+
+	mutex_enter(&bpl->bpl_lock);
+	while ((bpe = list_head(&bpl->bpl_list))) {
+		bplist_iterate_last_removed = bpe;
```
Comment (kithrup) [inline module/zfs/bplist.c]: I think that can go. Looking at the code, I didn't see anything that *used* bplist_iterate_last_removed, other than to assign to it.

File: module/zfs/dsl_dataset.c [hunk 15]
```diff
@@ -1212,6 +1241,26 @@ dsl_dataset_create_sync(dsl_dir_t *pdd, const char *lastname,
...
+	 * add the entry DD_FIELD_LIVELIST to ZAP.
+	 */
+	if (origin != NULL &&
+	    spa_feature_is_enabled(dp->dp_spa, SPA_FEATURE_LIVELIST)) {
+		/*
+		 * Livelists are not created for filesystems based on the
+		 * origin snap since they're not actually clones and
+		 * wouldn't benefit from using livelists.
+		 */
+		ASSERT3P(origin, !=, dp->dp_origin_snap);
```
Comment (kithrup) [inline module/zfs/dsl_dataset.c]: Can there be a case where origin != NULL but it is NOT dp->dp_origin_snap?  If not (which the assert would lead me to believe), why is the if condition above not for origin == dp->dp_origin_snap?

**Comment by shartse** [comment]
Thanks for taking a look! I'll go through you comments soon.

**Comment by behlendorf** [comment]
@shartse can you please rebase this on master now that redacted send/recv has been merged.

File: man/man5/zpool-features.5 [hunk 16]
```diff
@@ -884,5 +884,25 @@ The feature will only return back to being \fBenabled\fR when the pool
...
+READ\-ONLY COMPATIBLE	yes
+DEPENDENCIES	none
+.TE
+
+This feature allows clones to be deleted faster than the traditional method
+when there are large number of random/sparse writes made to the clone.
+All blocks allocated and freed after a clone is created are tracked by the
+the clone's livelist which is referenced during the deletion of the clone.
+The feature is activated when a clone is created and remains active until all
+clones have been destroyed.
```
Comment (behlendorf) [inline man/man5/zpool-features.5]: Rather than adding this to the end can you place it in alphabetical order with the other feature flags.

File: man/man5/zpool-features.5 [hunk 17]
```diff
@@ -884,5 +884,25 @@ The feature will only return back to being \fBenabled\fR when the pool
...
+.RS 4n
+.TS
+l l .
+GUID	com.delphix:livelist
+READ\-ONLY COMPATIBLE	yes
+DEPENDENCIES	none
+.TE
+
+This feature allows clones to be deleted faster than the traditional method
+when there are large number of random/sparse writes made to the clone.
```
Comment (behlendorf) [inline man/man5/zpool-features.5]: How about: "This feature allows clones to be deleted faster than the traditional method
when a large number of random/sparse writes have been made to the clone."

File: module/zfs/dsl_deadlist.c [hunk 18]
```diff
@@ -51,6 +51,68 @@
...
+ * to track blkptrs over the lifetime of a dataset. Livelists however, belong
+ * to clones and track the blkptrs that are clone-specific (were born after
+ * the clone's creation). The exception is embedded block pointers which are
+ * not included in Livelists because they do not need to be freed.
+ *
+ * When it comes time to delete the clone, the Livelist provides a quick
+ * reference as to what needs to be freed. For this reason, Livelists also track
+ * when clone-specific blkptrs are freed before deletion to prevent double
+ * frees. Each blkptr in a Livelist is marked as a FREE or an ALLOC and the
+ * deletion algorithm iterates backwards over the livelist, matching
```
[the remark this answers is not in the document]
Comment (shartse) [reply module/zfs/dsl_deadlist.c]: Changed to lower except for beginning of the sentences

File: module/zfs/dsl_deadlist.c [hunk 19]
```diff
@@ -572,3 +771,118 @@ dsl_deadlist_move_bpobj(dsl_deadlist_t *dl, bpobj_t *bpo, uint64_t mintxg,
...
+	const blkptr_t *r = ((livelist_entry_t *)rarg)->le_bp;
+
+	/* Sort them according to dva[0] */
+	uint64_t l_dva0_vdev, r_dva0_vdev;
+	l_dva0_vdev = DVA_GET_VDEV(&l->blk_dva[0]);
+	r_dva0_vdev = DVA_GET_VDEV(&r->blk_dva[0]);
+	if (l_dva0_vdev < r_dva0_vdev)
+		return (-1);
+	else if (l_dva0_vdev > r_dva0_vdev)
+		return (+1);
```
[the remark this answers is not in the document]
Comment (shartse) [reply module/zfs/dsl_deadlist.c]: Sounds like a good idea, though I guess this comparison is a little more complex different since it's based on two factors (and we have that assertion that entries at the same location have the same birth.) 

Here's what I changed it to - let me know if you have anything more efficient in mind. 

```
793         if (l_dva0_vdev != r_dva0_vdev)
794                 return (AVL_CMP(l_dva0_vdev, r_dva0_vdev));   
            ...
801         if (l_dva0_offset == r_dva0_offset)
802                 ASSERT3U(l->blk_birth, ==, r->blk_birth);
804         return (AVL_CMP(l_dva0_offset, r_dva0_offset));
```

File: module/zfs/dsl_deadlist.c (hunk 6 above)
Comment (shartse) [reply module/zfs/dsl_deadlist.c]: All exposed parameters are now documented in zfs-module-parameters.5

File: module/zfs/dsl_dir.c [hunk 20]
```diff
@@ -2224,6 +2242,86 @@ dsl_dir_is_zapified(dsl_dir_t *dd)
...
+				dmu_buf_rele(spa->spa_to_condense.ds->ds_dbuf,
+				    spa);
+				spa->spa_to_condense.ds = NULL;
+			}
+	}
+
+	dsl_dir_livelist_close(dd);
+	int err = zap_lookup(dp->dp_meta_objset, dd->dd_object,
+	    DD_FIELD_LIVELIST, sizeof (uint64_t), 1, &obj);
+	ASSERT3U(err, !=, ENOENT);
```
[the remark this answers is not in the document]
Comment (shartse) [reply module/zfs/dsl_dir.c]: For the `ECKSUM` handling - would you recommend just making sure the callers are protected (bail out it err != 0) or do you think I should make changes to propagate the `ECKSUM` error further up the stack?

File: module/zfs/bplist.c (hunk 14 above)
Comment (shartse) [reply module/zfs/bplist.c]: There's a comment above this saying that it's used for debugging - 
```
 /*
   * To aid debugging, we keep the most recently removed entry.  This way if
   * we are in the callback, we can easily locate the entry.
   */
```
so I think I'll leave it in for now.

File: module/zfs/spa.c [hunk 21]
```diff
@@ -2273,6 +2307,356 @@ spa_vdev_err(vdev_t *vdev, vdev_aux_t aux, int err)
...
+			dsl_sync_task_nowait(spa_get_dsl(spa),
+			    spa_livelist_condense_sync, lca, 0,
+			    ZFS_SPACE_CHECK_NONE, tx);
+			dmu_tx_commit(tx);
+			return;
+		}
+	}
+	/*
+	 * Condensing can not continue: either it was externally stopped or
+	 * the pool has run out of space
```
[the remark this answers is not in the document]
Comment (shartse) [reply module/zfs/spa.c]: How does this sound:
```
2677         /*
2678          * Condensing can not continue: either it was externally stopped or
2679          * we were unable to assign to a tx because the pool has run out of
2680          * space. In the second case, we'll just end up trying to condense again 
2681          * in a later txg.                                                                                                            
2682          */
```

File: module/zfs/dsl_deadlist.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (behlendorf) [reply module/zfs/dsl_deadlist.c]: That's nice.  My only suggestion would be replace the second conditional for the ASSERT with an IMPLY.  Though, it likely doesn't matter since I'd expect the compiler to entirely optimize out that code out for non-debug builds.  So I'm happy with it either way.  Your version does have the advantage that it will log the bad blk_birth values.

```diff
-         if (l_dva0_offset == r_dva0_offset)
-                 ASSERT3U(l->blk_birth, ==, r->blk_birth);
+         IMPLY(l_dva0_offset == r_dva0_offset, l->blk_birth == r->blk_birth);
```

File: module/zfs/dsl_dir.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (behlendorf) [reply module/zfs/dsl_dir.c]: The way we've handled this elsewhere is to return an error up the stack whenever we can.  This is probably the most important for the import path and should be pretty straight forward.  The ZTS extreme rewind are expected to encounter this periodically.  For the other places where we can't bail, if there's a failure that can't be handled without causing pool damage then there's a VERIFY.  Which isn't ideal, but addressing that is a project in and of itself.

[the remark this answers is not in the document]
Comment (shartse) [reply module/zfs/dsl_dir.c]: Cool - thanks for the suggestions. I'll add the error return in `dsl_dir_hold_obj` (which is what you mean by the import path, right?)

[the remark this answers is not in the document]
Comment (behlendorf) [reply module/zfs/dsl_dir.c]: For the import path I was worried about `spa_ld_get_props()`, but everything looks good there now!  For `dsl_dir_hold_obj()` actually, I think it may be fine.  What would be the consequence of not calling `dsl_dir_livelist_open()` here?

[the remark this answers is not in the document]
Comment (shartse) [reply module/zfs/dsl_dir.c]: It would just not use the livelist feature, which isn't a big deal. Though if it were to later successfully open the livelist, there would be issues will some ALLOCs not being tracked.

**Comment by shartse** [comment]
@behlendorf - Great thanks for you feedback! I updated this review with one more fix to address a race condition we found recently. It'd be great if you could take a look at that as well and let me know if you have any concerns.

File: module/zfs/dsl_dataset.c (hunk 15 above)
Comment (shartse) [reply module/zfs/dsl_dataset.c]: It took me awhile to figure this out (I'd totally forgotten where that assertion came from in the first place..) but basically it has to do with how the function is invoked and the implication that if we are creating a filesystem (a non-clone), we won't have the origin passed as equal to `dp->dp_origin_snap` (even though it eventually gets set to that). 

I think the point of having there assertion there was to firm up the assumption that anything with a non-NULL origin really is a clone. However, It's kind of complicated and non obvious. 

I've moved it to outside the clone case and added a more general comment that I think is easier to understand.

File: module/zfs/dsl_dir.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (shartse) [reply module/zfs/dsl_dir.c]: Brian and I talked offline about this and agreed that returning the err in this function is sufficient to prevent this livelist non-determinism possibility.

File: module/zfs/dsl_destroy.c [hunk 22]
```diff
@@ -852,6 +855,125 @@ dsl_dir_destroy_sync(uint64_t ddobj, dmu_tx_t *tx)
...
+	    DMU_POOL_DELETED_CLONES, sizeof (uint64_t), 1, &zap_obj);
+	if (error == ENOENT) {
+		zap_obj = zap_create(mos, DMU_OTN_ZAP_METADATA,
+		    DMU_OT_NONE, 0, tx);
+		VERIFY0(zap_add(mos, DMU_POOL_DIRECTORY_OBJECT,
+		    DMU_POOL_DELETED_CLONES, sizeof (uint64_t), 1,
+		    &(zap_obj), tx));
+		spa->spa_livelists_to_delete = zap_obj;
+	} else {
+		ASSERT0(error);
```
[the remark this answers is not in the document]
Comment (shartse) [reply module/zfs/dsl_destroy.c]: I talked to @ahrens about this and he suggested adding a call to `zfs_panic_recover` if we get corruption here.

**Comment by rrygl** [comment]
Can someone look at this corruption bug?
https://github.com/openzfs/zfs/issues/11480
It's believed to be caused by this new feature.
